File: flutter/android/src/main/kotlin/com/dkmads/ssp/flutter/DkmadsBannerPlatformView.kt
package com.dkmads.ssp.flutter

import android.content.Context
import android.view.View
import com.dkmads.ssp.DKMadsBannerAdView
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class DkmadsBannerViewFactory : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
  override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
    @Suppress("UNCHECKED_CAST")
    val params = args as? Map<String, Any?> ?: emptyMap()
    return DkmadsBannerPlatformView(context, params)
  }
}

class DkmadsBannerPlatformView(
  context: Context,
  params: Map<String, Any?>,
) : PlatformView {
  private val banner = DKMadsBannerAdView(
    context = context,
    adUnitId = params["adUnitId"] as? String ?: "",
    adWidth = (params["width"] as? Number)?.toInt() ?: 300,
    adHeight = (params["height"] as? Number)?.toInt() ?: 250,
  )

  init {
    if (params["autoLoad"] == true) {
      banner.load()
    }
  }

  override fun getView(): View = banner

  override fun dispose() {
    banner.destroy()
  }
}

File: flutter/ios/Classes/DkmadsBannerPlatformView.swift
import Flutter
import UIKit

final class DkmadsBannerPlatformView: NSObject, FlutterPlatformView {
  private let banner: DKMadsBannerAdView

  init(frame: CGRect, viewId: Int64, args: [String: Any]?) {
    let adUnitId = args?["adUnitId"] as? String ?? ""
    let w = (args?["width"] as? NSNumber)?.doubleValue ?? Double(frame.width)
    let h = (args?["height"] as? NSNumber)?.doubleValue ?? Double(frame.height)
    let width = w > 0 ? w : 300
    let height = h > 0 ? h : 250
    banner = DKMadsBannerAdView(adUnitID: adUnitId, adSize: CGSize(width: width, height: height))
    banner.translatesAutoresizingMaskIntoConstraints = false
    super.init()
    if frame.width > 0, frame.height > 0 {
      banner.setAdSize(frame.size)
    }
    if let auto = args?["autoLoad"] as? Bool, auto {
      banner.load()
    }
  }

  func view() -> UIView { banner }
}

final class DkmadsBannerViewFactory: NSObject, FlutterPlatformViewFactory {
  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    let params = args as? [String: Any] ?? [:]
    return DkmadsBannerPlatformView(frame: frame, viewId: viewId, args: params)
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    FlutterStandardMessageCodec.sharedInstance()
  }
}
